Mathlib/Order/Grade.lean

```diff
@@ -0,0 +1,368 @@
+* `grade`: The grade of an element. Because an order can admit several gradings, the first argument
+  is the order we grade by.
+
+## How to grade your order
+
+Here are the translations between common references and our `GradeOrder`:
+* [Stanley][stanley2012] defines a graded order of rank `n` as an order where all maximal chains
+  have "length" `n` (so the number of elements of a chain is `n + 1`). This corresponds to
+  `GradeBoundedOrder (Fin (n + 1)) α`.
+* [Engel][engel1997]'s ranked orders are somewhere between `GradeOrder ℕ α` and
+  `GradeMinOrder ℕ α`, in that he requires `∃ a, IsMin a ∧ grade ℕ a = 0` rather than
+  `∀ a, IsMin a → grade ℕ a = 0`. He defines a graded order as an order where all minimal elements
+  have grade `0` and all maximal elements have the same grade. This is roughly a less bundled
+  version of `GradeBoundedOrder (Fin n) α`, assuming we discard orders with infinite chains.
+
+## Implementation notes
+
+One possible definition of graded orders is as the bounded orders whose flags (maximal chains)
+all have the same finite length (see Stanley p. 99). However, this means that all graded orders must
+have minimal and maximal elements and that the grade is not data.
+
+Instead, we define graded orders by their grade function, without talking about flags yet.
+
+## References
+
+* [Konrad Engel, *Sperner Theory*][engel1997]
+* [Richard Stanley, *Enumerative Combinatorics*][stanley2012]
+-/
+
+open Finset Nat OrderDual
+
+variable {𝕆 ℙ α β : Type _}
+
+/-- An `𝕆`-graded order is an order `α` equipped with a strictly monotone function
+`grade 𝕆 : α → 𝕆` which preserves order covering (`Covby`). -/
+class GradeOrder (𝕆 α : Type _) [Preorder 𝕆] [Preorder α] where
+  /-- The grading function. -/
+  grade : α → 𝕆
+  /-- `grade` is strictly monotonic. -/
+  grade_strict_mono : StrictMono grade
+  /-- `grade` preserves `Covby`. -/
+  covby_grade ⦃a b : α⦄ : a ⋖ b → grade a ⋖ grade b
+#align grade_order GradeOrder
+
+/-- A `𝕆`-graded order where minimal elements have minimal grades. -/
+class GradeMinOrder (𝕆 α : Type _) [Preorder 𝕆] [Preorder α] extends GradeOrder 𝕆 α where
+  /-- Minimal elements have minimal grades. -/
+  is_min_grade ⦃a : α⦄ : IsMin a → IsMin (grade a)
+#align grade_min_order GradeMinOrder
+
+/-- A `𝕆`-graded order where maximal elements have maximal grades. -/
+class GradeMaxOrder (𝕆 α : Type _) [Preorder 𝕆] [Preorder α] extends GradeOrder 𝕆 α where
+  /-- Maximal elements have maximal grades. -/
+  is_max_grade ⦃a : α⦄ : IsMax a → IsMax (grade a)
+#align grade_max_order GradeMaxOrder
+
+/-- A `𝕆`-graded order where minimal elements have minimal grades and maximal elements have maximal
+grades. -/
+class GradeBoundedOrder (𝕆 α : Type _) [Preorder 𝕆] [Preorder α] extends GradeMinOrder 𝕆 α,
+  GradeMaxOrder 𝕆 α
+#align grade_bounded_order GradeBoundedOrder
+
+section Preorder -- grading
+-- PORTING NOTE: this `variable [Preorder 𝕆]` for the whole section seems to not work in Lean4
+variable [Preorder 𝕆]
```

Comment:
Maybe it's because of autoimplicits? In any case, I think I would comment out this line, as part of the porting note. To make it clear that you've added the variable back in many other places.
```suggestion
-- PORTING NOTE: this `variable [Preorder 𝕆]` for the whole section seems to not work in Lean4
-- variable [Preorder 𝕆]
```